### PR TITLE
storage: use buckets in BoltDB backend

### PR DESCRIPTION
**Pull request by fyrchik**
1. Splitting 1-byte prefix _seems_ to increas restore time by about  25% under heavy workload. However this difference is actually flat 25s gained only during first persist cycle (on master branch it takes 25s, while here it takes 1s). When restoring is done from the 10-th block, running times don't differ at all. Same thing with `KeepOnlyLatestState: true`.
2. Single-node benchmark performance difference is negligible.

Keys in bucket aren't stripped because it is prohibited to put empty key in bucket. With stripping implementation will become more complicate and I expect no performance gains because of the need to append prefix in `Seek`.

Some other bucketing strategies can be discussed, however it is obvious that most of the load is due to MPT.

Stats regarding keys in DB for this bench (100 blocks, 10k tx per block, single `storage.Put` for the same contract per tx):
1. `DataMPT` (stores MPT nodes) contains ~4M keys. Splitting this into another 256 buckets decreases B-tree depth from 4 to 3. The distribution is uniform because keys are hashes, so this makes sense.
2. `DataTransaction` and `STStorage` contain ~1M keys.
3. Other prefixes contain much less keys.

TODO:
1. Extend dump generation to include random NEO + GAS transfers.
2. Check if creating separate buckets for frequently used contracts (NEO, GAS) affects running time.
3. Use _sub_ buckets for MPT.

### CODE REVIEW

**Comment by fyrchik** [comment]
Moving NEO/GAS contract storage to a separate bucket also doesn't affect restore time.

File: pkg/core/storage/boltdb_store.go [hunk 1]
```diff
@@ -78,23 +94,25 @@ func (s *BoltDBStore) Get(key []byte) (val []byte, err error) {
...
-		return b.Delete(key)
+		k1, k2 := split(key)
+		b := tx.Bucket(k1)
+		return b.Delete(k2)
 	})
 }
 
 // PutBatch implements the Store interface.
 func (s *BoltDBStore) PutBatch(batch Batch) error {
 	return s.db.Batch(func(tx *bbolt.Tx) error {
```
Comment (roman-khimov) [inline pkg/core/storage/boltdb_store.go:105]: And we can use `Update` here, I think.

File: pkg/core/storage/boltdb_store.go [hunk 2]
```diff
@@ -37,30 +33,50 @@ func NewBoltDBStore(cfg BoltDBOptions) (*BoltDBStore, error) {
...
+
+func split(key []byte) ([]byte, []byte) {
+	switch KeyPrefix(key[0]) {
+	case DataMPT:
+		if len(key) >= 2 {
+			return key[:2], key
+		}
+		return key[:1], key
+	default:
+		return key[:1], key
```
Comment (roman-khimov) [inline pkg/core/storage/boltdb_store.go:60]: We have very limited number of cases where `key[1:]` is zero-length, maybe we should try fixing them and providing proper split here. I'm not sure it'll change anything, but who knows.

File: pkg/core/storage/boltdb_store.go [hunk 3]
```diff
@@ -37,30 +33,50 @@ func NewBoltDBStore(cfg BoltDBOptions) (*BoltDBStore, error) {
...
 	err = db.Update(func(tx *bbolt.Tx) error {
-		_, err = tx.CreateBucketIfNotExists(Bucket)
-		if err != nil {
-			return fmt.Errorf("could not create root bucket: %w", err)
+		for i := 0; i <= 255; i++ {
+			_, err := tx.CreateBucketIfNotExists([]byte{byte(i)})
+			if err != nil {
+				return err
+			}
+			_, err = tx.CreateBucketIfNotExists([]byte{byte(DataMPT), byte(i)})
```
Comment (roman-khimov) [inline pkg/core/storage/boltdb_store.go:41]: MPT probably is not worth splitting.

**Review by roman-khimov — COMMENTED** [review]
Have you tried more blocks (1000)? And changing block/tx ratio (1K per block, 10K blocks for example)?

**Comment by roman-khimov** [comment]
OK, I've played with it a little and no matter what I do nothing really changes, it's all at the level of test tolerance. So just adding buckets doesn't magically improve performance and thus it's not worth doing.

**Pull request closed without merging**